src/linear_algebra/basis.lean

```diff
@@ -82,6 +82,23 @@ theorem linear_independent_iff : linear_independent R v ↔
   ∀l, finsupp.total ι M R v l = 0 → l = 0 :=
 by simp [linear_independent, linear_map.ker_eq_bot']
 
+theorem linear_independent_iff' : linear_independent R v ↔
+  ∀ s : finset ι, ∀ g : ι → R, s.sum (λ i, g i • v i) = 0 → ∀ i ∈ s, g i = 0 :=
+by classical; exact linear_independent_iff.trans
+⟨λ hf s g hg, have h : _ := hf ⟨s.filter (λ i, g i ≠ 0), λ i, if i ∈ s then g i else 0,
+    λ i, by simp only [finset.mem_filter]; split_ifs; cc⟩ $ calc
```

Comment:
This proof depends heavily on the representation of `finsupp`. How about starting the argument like
```lean
⟨λ hf s g hg,
  have h : _ := hf (s.sum (λ i, finsupp.single i (g i)))
    (by simpa only [linear_map.map_sum, finsupp.total_single] using hg),
  assume i hi, _,
```
(The rest of the proof could be
```lean
  begin
    convert ←(congr_arg (λ (x : ι →₀ R), x i) h),
    apply (finsupp.lapply i : (ι →₀ R) →ₗ[R] R).map_sum.trans,
    simp only [finsupp.single_apply, finsupp.lapply_apply],
    convert finset.sum_ite s i (g i),
    { ext x,
      conv { to_lhs, rw eq_comm },
      split_ifs with hx,
      simp [hx] },
    { simp [hi] }
  end,
```
but I imagine you could come up with something better.)

Reply:
I think the equality test in `finsupp.single_apply` should be written the other way, which would help a bit.

Reply:
done